crawler/crawl_emap.py

```diff
@@ -0,0 +1,130 @@
+            retries = 0
+            while True:
+                try:
+                    result_func = func(*args, **kwargs)
+                    return result_func
+                except Exception as ex:
+                    current_app.logger.debug(f"Attempt {retries + 1} failed: {ex}")
+                    if retries < max_retries and should_retry(ex):
+                        delay = (base_delay_s * 2 ** retries + random.uniform(0, 1))
+                        current_app.logger.debug(f"Retrying in {delay:.2f} seconds...")
+                        sleep(delay)
+                        retries += 1
+                    else:
+                        raise
+        return wrapper
+    return decorator
+
+def get_auth_token():
+    try:
+        parser = configparser.ConfigParser()
+        config_filepath = CONFIG_FILE
+        parser.read(config_filepath)
+        return parser['auth']['token']
+    except Exception as ex:
+        raise ValueError(
+            f'Failed to retrieve watttime credentials: {ex}') from ex
+
+
+def prepare_insert_query(data):
+    carbon_intensity = data.get('carbonIntensity', 0.0)
+    if carbon_intensity is None:
+        carbon_intensity = 0.0
+
+    low_carbon_percentage = 0.00
+    renewable_percentage = 0.00
+
+    datetime = data['datetime']
+
+    zone_id = data['zone']
+
+    query = """
+    INSERT INTO EMapCarbonIntensity(DateTime, ZoneId, CarbonIntensity, LowCarbonPercentage, RenewablePercentage)
+    VALUES (%s, %s, %s, %s, %s)
+    ON CONFLICT (DateTime, ZoneId) DO UPDATE SET
+    CarbonIntensity = EXCLUDED.CarbonIntensity,
+    LowCarbonPercentage = EXCLUDED.LowCarbonPercentage,
+    RenewablePercentage = EXCLUDED.RenewablePercentage;
+    """
+
+    return query, (datetime, zone_id, carbon_intensity, low_carbon_percentage, renewable_percentage)
+
+@exponential_backoff(should_retry=lambda ex: ex.response.status_code == 429)
+def fetch(zone):
+    url = f"https://api-access.electricitymaps.com/free-tier/carbon-intensity/history?zone={zone}"
+    headers = {
+        "auth-token": get_auth_token()
+    }
+
+    history_response = requests.get(url, headers=headers)
+
+    history_response_json = history_response.json()
+    try:
+        assert history_response.ok, f"Error in EMap request for getting history according to zone ({history_response.status_code}): {history_response.text}"
+        # Rest of your code goes here, which will execute if the assertion passes
+    except AssertionError as error:
+        print(str(error), file=sys.stderr)
+
+    return history_response_json
+
+
+def update(conn, history_response_json):
+    if 'history' in history_response_json:
+        for history in history_response_json['history']:
+            query, data = prepare_insert_query(history)
+            with conn, conn.cursor() as cur:
+                try:
+                    cur.execute(query, data)
+                except psycopg2.Error as ex:
+                    print(datetime.now().isoformat(),
```

Comment:
Please also print the stack trace. Check the crawl.py for examples (`traceback.format_exc()` iirc).

Reply:
Actually it's better to throw an exception here saying failed to execute and let the outer function to catch this user-thrown exception and log it, so your error logging code is in one place. Essentially fetch can throw an error, update can throw another error, either error should stop the code execution of this region, and if that happens, we report the error and move on.

Reply:
fixed, now, the outer function "fetch and update" will catch this user-thrown exception.